src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,337 @@
+                    if (_symbolValueUsageStateMap.TryGetValue(memberSymbol, out var currentUsageInfo))
+                    {
+                        usageInfo = currentUsageInfo | usageInfo;
+                    }
+
+                    _symbolValueUsageStateMap[memberSymbol] = usageInfo;
+                }
+            }
+
+            private bool TryRemove(ISymbol memberSymbol, out ValueUsageInfo valueUsageInfo)
+            {
+                lock (_gate)
+                {
+                    if (_symbolValueUsageStateMap.TryGetValue(memberSymbol, out valueUsageInfo))
+                    {
+                        _symbolValueUsageStateMap.Remove(memberSymbol);
+                        return true;
+                    }
+
+                    return false;
+                }
+            }
+
+            public void AnalyzeMemberReferenceOperation(OperationAnalysisContext operationContext)
+            {
+                var memberReference = (IMemberReferenceOperation)operationContext.Operation;
+                if (IsCandidateSymbol(memberReference.Member))
+                {
+                    // Get the value usage info.
+                    var valueUsageInfo = memberReference.GetValueUsageInfo();
+
+                    if (valueUsageInfo == ValueUsageInfo.ReadWrite)
+                    {
+                        Debug.Assert(memberReference.Parent is ICompoundAssignmentOperation compoundAssignment &&
+                            compoundAssignment.Target == memberReference ||
+                            memberReference.Parent is IIncrementOrDecrementOperation);
+
+                        // Compound assignment or increment whose value is being dropped (parent has null type)
+                        // is treated as a Write as the value was never actually 'read' in a way that is observable.
+                        if (memberReference.Parent.Parent?.Type == null)
+                        {
+                            valueUsageInfo = ValueUsageInfo.Write;
+                        }
+                    }
+
+                    OnSymbolUsage(memberReference.Member, valueUsageInfo);
+                }
+            }
+
+            public void AnalyzeInvocationOperation(OperationAnalysisContext operationContext)
+            {
+                var invocation = (IInvocationOperation)operationContext.Operation;
+                if (IsCandidateSymbol(invocation.TargetMethod))
+                {
+                    OnSymbolUsage(invocation.TargetMethod, ValueUsageInfo.Read);
+                }
+            }
+
+            public void OnSymbolEnd(SymbolAnalysisContext symbolEndContext, bool hasInvalidOperation)
+            {
+                if (hasInvalidOperation)
+                {
+                    return;
+                }
+
+                // Report diagnostics for unused candidate members.
+                ImmutableHashSet<ISymbol> symbolsReferencedInDocComments = null;
```

Comment:
in general, making these for ephemeral computed values is fairly expensive.  consider just a hashset.  Also, consider pooling.